Kokkos::SIMD::SVE support for 128-bit and 256-bit SVE

hominhquan opened the pull request:
- This work is a continuation of a previous (incomplete) PR #7109.
  The original commit of #7109 was rebased over the `develop` branch,
  before being integrated into this PR and followed by new modification.
- Many compilation issues (related to SVE intrinsics syntax and missing
  operators) were resolved, as well as correctness errors (e.g. in
  gather-loads, inactive lanes should be left unchanged, whereas they are
  zero-ed out in SVE; semantic mismatch in implementation of `operator=`).

- Support any ARMv8.4-a+SVE processors, with 128-bit and 256-bit SVE VL
  (except 512-bit due to current implementation limitation).
- Auto-detect SVE VL size at configure (CMake try_run()) on the
  running CPU and insert `-msve-vector-bits=<VL>` to compile flag.
- Enabled with KOKKOS_ARCH_NATIVE, KOKKOS_ARCH_ARMV84_SVE and
  KOKKOS_ARCH_ARMV9_GRACE.

- simd/unit_tests: implementation of a simd<T, Abi> requires implicitly the
  availability of simd_mask<T, Abi> for comparison operators (<,>,=).
  The boolean tester 'is_type_v<>' is now extended to 'is_simd_avail_v<T, Abi>'
  to check for both simd<T, Abi> and simd_mask<T, Abi>.

crtrott commented:
Just FYI, we have been talking about this and this is under review. Thank you so much for the contribution we appreciate it, and definitely want to improve support for ARM in general, and SiPearl specifically as part of our support for the European initiatives. Question: if we see small things that need updating do you like us to push your branch or just report it?

crtrott commented on core/perf_test/CMakeLists.txt:
```diff
@@ -28,15 +28,21 @@ find_package(benchmark QUIET 1.5.6)
+    set(BENCHMARK_SRC_URL ${CMAKE_SOURCE_DIR}/tpls/googlebenchmark-v1.7.1.tar.gz)
+  else()
+    set(BENCHMARK_SRC_URL https://github.com/google/benchmark/archive/refs/tags/v1.7.1.tar.gz)
+  endif()
```
what is this? It looks like you added locally a tar ball for google benchmark to Kokkos?

hominhquan replied:
yes, this is a small trick to avoid always (re)downloading from the internet each time.

Note: these test-related changes are put in a separate commit from the commit of SVE backend.

hominhquan replied:
This is already in another commit (2 commits in this PRs). Should it be in another PR instead ?

crtrott commented on simd/src/Kokkos_SIMD.hpp:
```diff
@@ -84,33 +95,45 @@ namespace simd_abi {
 
 #if defined(KOKKOS_ARCH_AVX512XEON)
+template <class T>
 using host_fixed_native = avx512_fixed_size<8>;
```
I don't think I agree with this right now. We may need to rethink this alltogether but this is very specific choices based on the number of `double` values fitting into a simd register.

hominhquan replied:
Please see comment https://github.com/kokkos/kokkos/pull/7807#discussion_r1979528085

crtrott commented on simd/src/Kokkos_SIMD.hpp:
```diff
@@ -84,33 +95,45 @@ namespace simd_abi {
+#elif defined(KOKKOS_ARCH_ARM_SVE)
+template <class T>
+using host_fixed_native =
+    sve_fixed_size<(__ARM_FEATURE_SVE_BITS / (8 * sizeof(T)))>;
```
I am pretty sure our previous use of these host_fixed_native thingies was so that a SIMD of double and SIMD of float etc. have all the same size (number of elements). I.e. that you could in principle multiple simd of double and simd of float or at least assign them to each other. In that spirit I think the template needs to go and `host_fixed_native = see_fixed_size<(__ARM_FEATURE_SVE_BITS/sizeof(double))>`

@ldh4 is my understanding correct?

crtrott reviewed (changes requested):
I did an initial glance through and noted a few high-level things as well as made comments on some of the pre-existing files changed.

hominhquan commented on benchmarks/gups/gups.cpp:
```diff
@@ -123,7 +123,7 @@ int run_benchmark(const Index indicesCount, const Index dataCount,
       "init-data",
       Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(0, dataCount),
-      KOKKOS_LAMBDA(const int i) { data[i] = 10101010101; });
+      KOKKOS_LAMBDA(const Index i) { data[i] = 10101010101; });
```
this change allows running the benchmark with size larger than `INT_MAX`

hominhquan replied:
Note: This change is put in a second commit (65347168) after the first one on SVE backend.

By "change the IndexType for the RangePolicy" did you mean `using Index = int64_t;` at line L33 ?

masterleinad replied:
No, you would need
```
Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace, Kokkos::IndexType<Index>>(
```
to use that index type also internally (instead of the default which is backend-specific but typically `int` or `unsigned int`).

hominhquan commented on cmake/kokkos_arch.cmake:
```diff
@@ -345,10 +417,11 @@ endif()
 
 if(KOKKOS_ARCH_ARMV9_GRACE)
   set(KOKKOS_ARCH_ARM_NEON ON)
-  check_cxx_compiler_flag("-mcpu=neoverse-n2" COMPILER_SUPPORTS_NEOVERSE_N2)
```
This is what I was confused about. Grace CPU is built with [Neoverse-v2 cores](https://docs.nvidia.com/grace-perf-tuning-guide/index.html), and here is set to `neoverse-n2` here. If this is really what NVIDIA meant to be, I'll rollback this change. please confirm ?

cedricchevalier19 replied:
I think too that `v2` is more appropriate.

hominhquan commented:
> Just FYI, we have been talking about this and this is under review. Thank you so much for the contribution we appreciate it, and definitely want to improve support for ARM in general, and SiPearl specifically as part of our support for the European initiatives. 

Yes, we are open to collaboration.

> Question: if we see small things that need updating do you like us to push your branch or just report it?

Please report any remark and comment directly in github review, I'll update my branch with new changes.

ldh4 reviewed (commented):
You may need to rebase once again. I think a recently merged changes on simd comparison operators will cause a compilation error.

dssgabriel commented on Makefile.kokkos:
```diff
@@ -812,11 +827,29 @@ endif
+                                                     && $(tmp) | sed 's/SVE_HW_VL=//' \
+                                                     ; rm -f $(tmp) ))
+
+  KOKKOS_CXXFLAGS += -march=armv8.4-a+sve -msve-vector-bits=$(KOKKOS_INTERNAL_ARCH_ARM_SVE_VL)
+  KOKKOS_LDFLAGS += -march=armv8.4-a+sve -msve-vector-bits=$(KOKKOS_INTERNAL_ARCH_ARM_SVE_VL)
```
I am not aware of any Armv8.4-A CPU with SVE support that uses a vector width different than 256 (because these are basically Neoverse V1-based, anyway). Maybe this could be simplified as:
```suggestion
  KOKKOS_CXXFLAGS += -march=armv8.4-a+sve -msve-vector-bits=256
  KOKKOS_LDFLAGS += -march=armv8.4-a+sve -msve-vector-bits=256
```

Alternatively, it might be simpler to redefine `Kokkos_ARCH_ARMV84_SVE` as `Kokkos_ARCH_NEOVERSE_V1`to use an Arm _microarchitecture_ instead of a generic architecture. It is kind of what Kokkos already does with, e.g., A64FX (which is its own microarchitecture implementing Armv8.2-A+SVE).

hominhquan replied:
I agree that Neoverse-V1 is today the only one in Armv8.4-a+SVE(1) config, and SVE1 technically does not prohibit any 512, 1024 or 2048-bit.

In this PR, we aimed to support any Armv8.4-a+SVE (not necessarily Neoverse-V1). This VL-detection mechanism is used for both Grace (128-bit) and Graviton3 (256-bit), and can be used for A64FX, too, as it does not cost much to check for the real HW SVE VL at configure.

dssgabriel commented on cmake/kokkos_arch.cmake:
```diff
@@ -48,6 +48,8 @@ declare_and_check_host_arch(NATIVE "local machine")
 declare_and_check_host_arch(ARMV80 "ARMv8.0 Compatible CPU")
 declare_and_check_host_arch(ARMV81 "ARMv8.1 Compatible CPU")
+declare_and_check_host_arch(ARMV84 "ARMv8.4 Compatible CPU")
+declare_and_check_host_arch(ARMV84_SVE "Generic ARMv8.4 with SVE support (-march=armv8.4-a+sve)")
```
Given that this is basically the architecture of a Neoverse V1-based CPU, it may be simpler to define this as `NEOVERSE_V1`, i.e., use a specific Arm _microarchitecture_ instead of a generic architecture. It is kind of what Kokkos already does with, e.g., A64FX (which is its own microarchitecture implementing Armv8.2-A+SVE). As far as I know, this would also better match SiPearl's Rhea1 (and AWS Graviton3).

hominhquan replied:
Honestly, I'm not keen on setting arch-specific CPU option for such a SVE ISA, which was aimed to be generic and portable. It would introduce further CPU-options bloating in Kokkos. We can already see it with `KOKKOS_ARCH_A64FX` and `KOKKOS_ARCH_ARMV9_GRACE`, which can typically use the same SVE backend implementation, with just different VL-fixed sizes. We can also imagine the scenario with future RISC-V architectures & vendors.

BTW, it remains my personal view, I'd like to hear other opinions on this "fragmentation-prone" aspect.

cedricchevalier19 replied:
Perhaps in the future we can have a look at https://github.com/archspec and try to be coherent with their model.

tretre91 commented on simd/src/Kokkos_SIMD.hpp:
```diff
@@ -253,6 +285,13 @@ using host_abi_set    = abi_set<simd_abi::scalar, simd_abi::avx2_fixed_size<4>,
 using data_type_set =
     data_types<std::int32_t, std::int64_t, std::uint64_t, double, float>;
+#elif defined(KOKKOS_ARCH_ARM_SVE)
+using host_abi_set =
+    abi_set<simd_abi::scalar, simd_abi::sve_fixed_size<2>,
+            simd_abi::sve_fixed_size<4>, simd_abi::sve_fixed_size<8>,
+            simd_abi::sve_fixed_size<16>>;
```
```suggestion
using host_abi_set =
    abi_set<simd_abi::scalar, simd_abi::sve_fixed_size<__ARM_FEATURE_SVE_BITS / 32>,
            simd_abi::sve_fixed_size<__ARM_FEATURE_SVE_BITS / 64>>;
```
If I'm not mistaken, depending on the SVE vector length there will always be two of those abis which would lead to compile errors if used, e.g. on Grace, `basic_simd<T, sve_fixed_size<8>>` and `basic_simd<T, sve_fixed_size<16>>` would be undefined.

hominhquan commented:
@ all : I've pushed a new update with following changes: 
- Rebased over `develop`
- Applied comments by @ldh4 about `protected`, `noexcept` and `friend`
- @crtrott: confirm about `-mcpu=neoverse-[n,v]2` for Grace ?
- @crtrott @ldh4 : confirm relevance of adding `template <class T>` to `host_fixed_native` ?
- Added internal implementation of two Neon-SVE bridge functions (`svget_neonq_u64()`, `svget_neonq_s32()`) for compilers which do not provide them (older than GCC14 and LLVM14).
- Add some more tests in [TestSIMD_Conversions.hpp](https://github.com/kokkos/kokkos/pull/7807/commits/9b2b01ac7dcf4019f952c944cc94daa40ae54d18#diff-d6f9b31253fd675f3f9a3d4203c4f8b729197127a626da8dabccf5b4c4f1fb2d) to check integer/boolean vectors of different values, rather than the current unique value of `1` or `true`.
- Passed the Grace-Grace CI with the new SVE backend (GCC13).
- Some clang-format errors on unrelated portion of code ( AMD GPU !?)

ldh4 reviewed (commented):
> Some clang-format errors on unrelated portion of code ( AMD GPU !?)

Ironic. Locally running clang-format 16 formats it back to what it is currently in `develop`.

hominhquan commented on Makefile.kokkos:
```diff
@@ -15,7 +15,7 @@ KOKKOS_DEVICES ?= "Threads"
 # Intel:    KNC,KNL,SNB,HSW,BDW,SKL,SKX,ICL,ICX,SPR
 # NVIDIA:   Kepler,Kepler30,Kepler32,Kepler35,Kepler37,Maxwell,Maxwell50,Maxwell52,Maxwell53,Pascal60,Pascal61,Volta70,Volta72,Turing75,Ampere80,Ampere86,Ada89,Hopper90
-# ARM:      ARMv80,ARMv81,ARMv8-ThunderX,ARMv8-TX2,A64FX,ARMv9-Grace
+# ARM:      ARMv80,ARMv81,ARMv8-ThunderX,ARMv8-TX2,A64FX,ARMv9-Grace,ARMV84,ARMV84_SVE
```
ok

hominhquan commented on simd/src/Kokkos_SIMD.hpp:
```diff
@@ -62,6 +66,13 @@
 #include <Kokkos_SIMD_AVX512.hpp>
 #endif
 
+#if defined(KOKKOS_ARCH_ARM_SVE)
+#if !defined(__ARM_FEATURE_SVE_BITS)
+#error "__ARM_FEATURE_SVE_BITS must be definded for KOKKOS_ARCH_ARM_SVE"
+#endif
+#include <Kokkos_SIMD_SVE.hpp>
+#endif
```
ok

hominhquan commented on simd/src/Kokkos_SIMD_SVE.hpp:
```diff
@@ -0,0 +1,3192 @@
+#define SVE_DOUBLES_IN_VECTOR (__ARM_FEATURE_SVE_BITS / 64)
+#define SVE_QUADS_IN_VECTOR (__ARM_FEATURE_SVE_BITS / 128)
+
+template <typename T>
+constexpr bool is_sve_vla =
+    (std::is_same_v<T, svint32_t> || std::is_same_v<T, svuint32_t> ||
+     std::is_same_v<T, svfloat32_t> || std::is_same_v<T, svint64_t> ||
+     std::is_same_v<T, svuint64_t> || std::is_same_v<T, svfloat64_t> ||
+     std::is_same_v<T, svbool_t>)
+        ? true
+        : false;
+
+template <typename T>
+constexpr bool is_sve_vls =
+    (std::is_same_v<T, vls_int32_t> || std::is_same_v<T, vls_uint32_t> ||
+     std::is_same_v<T, vls_float32_t> || std::is_same_v<T, vls_int64_t> ||
+     std::is_same_v<T, vls_uint64_t> || std::is_same_v<T, vls_float64_t> ||
+     std::is_same_v<T, vls_bool_t>)
+        ? true
+        : false;
```
Right, they were used during some development and debugging phases, but not used anymore. I'll remove them.

hominhquan commented:
> > Some clang-format errors on unrelated portion of code ( AMD GPU !?)
> 
> Ironic. Locally running clang-format 16 formats it back to what it is currently in `develop`.

BTW, it would be more pratical to run any clang-format version on local machines with a `clang-format.yml` config file, rather than using a fixed `clang-format-14` version on Github CI (which is quite outdated).

hominhquan commented on cmake/kokkos_arch.cmake:
```diff
@@ -255,7 +257,12 @@ if(KOKKOS_ARCH_NATIVE)
     set(KOKKOS_NATIVE_FLAGS "-mcpu=native")
   endif()
-  compiler_specific_flags(COMPILER_ID KOKKOS_CXX_HOST_COMPILER_ID NVHPC -tp=native DEFAULT ${KOKKOS_NATIVE_FLAGS})
+
+  if(KOKKOS_CXX_HOST_COMPILER_ID STREQUAL "NVHPC")
+    set(KOKKOS_NATIVE_FLAGS "-tp=native")
+  endif()
+
+  compiler_specific_flags(COMPILER_ID KOKKOS_CXX_HOST_COMPILER_ID DEFAULT ${KOKKOS_NATIVE_FLAGS})
```
This allows refactoring of `KOKKOS_ARCH_NATIVE` flags when SVE-enabled for all compilers at lines L771-781 of this file (`KOKKOS_NATIVE_FLAGS`).

ldh4 commented:
Other than the seemingly spurious clang-format issue, looks good to me.

ldh4 commented:
> Who else I should ping to get more approvals to be merged ?

Sorry for the delay. The clang-format check will need to be resolved in order to go through the rest of staged CI checks. I'm hoping to get this moving forward with #7919.

Meanwhile, you will need to apply `cmake-format` for changes in `cmake/kokkos_arch.cmake` file.

hominhquan commented:
> Sorry for the delay. The clang-format check will need to be resolved in order to go through the rest of staged CI checks. I'm hoping to get this moving forward with #7919.
> 
> Meanwhile, you will need to apply `cmake-format` for changes in `cmake/kokkos_arch.cmake` file.

FYI, I've just re-force-pushed a freshly rebased one, thanks.

hominhquan commented:
Re-pushed to add the new file `cmake/compile_tests/get_sve_hw_vl.cpp` into `scripts/diff_files`.

hominhquan commented:
@ldh4 CI job failed on SYCL-OneAPI config with the following error. I wonder if this is an environment issue ?
```
CMake Error at cmake/Modules/FindTPLONEDPL.cmake:34 (find_package):
  By not providing "FindoneDPL.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "oneDPL", but
  CMake did not find one.
```

ldh4 commented:
> @ldh4 CI job failed on SYCL-OneAPI config with the following error. I wonder if this is an environment issue ?
> 
> ```
> CMake Error at cmake/Modules/FindTPLONEDPL.cmake:34 (find_package):
>   By not providing "FindoneDPL.cmake" in CMAKE_MODULE_PATH this project has
>   asked CMake to find a package configuration file provided by "oneDPL", but
>   CMake did not find one.
> ```

The SYCL build failure isn't related to changes in this PR. But the `cmake-format-check` is failing on the modified `cmake/kokkos_arch.cmake` file:

<details>
<summary> Output: </summary>

```
diff --git a/cmake/kokkos_arch.cmake b/cmake/kokkos_arch.cmake
index 90b07e2..1650024 100644
--- a/cmake/kokkos_arch.cmake
+++ b/cmake/kokkos_arch.cmake
@@ -350,28 +350,32 @@ if(KOKKOS_ARCH_ARMV8_THUNDERX2)
 endif()
 
 # SVE helper function to query bitwise HW SVE length
-function (GET_SVE_HW_VL FLAG)
+function(GET_SVE_HW_VL FLAG)
   # if env var SVE_HW_VL is set, use it
   if(DEFINED ENV{SVE_HW_VL})
     set(SVE_HW_VL $ENV{SVE_HW_VL})
     message(STATUS "Using SVE_HW_VL from ENV{SVE_HW_VL} = ${SVE_HW_VL}")
   else()
-    try_run(RUN_GET_SVE_HW_VL COMPILE_GET_SVE_HW_VL
-          ${CMAKE_CURRENT_BINARY_DIR}
-          ${CMAKE_CURRENT_SOURCE_DIR}/cmake/compile_tests/get_sve_hw_vl.cpp
-          COMPILE_DEFINITIONS ${FLAG}
-          COMPILE_OUTPUT_VARIABLE COMPILE_OUTPUT_GET_SVE_HW_VL
-          RUN_OUTPUT_VARIABLE SVE_HW_VL)
+    try_run(
+      RUN_GET_SVE_HW_VL COMPILE_GET_SVE_HW_VL ${CMAKE_CURRENT_BINARY_DIR}
+      ${CMAKE_CURRENT_SOURCE_DIR}/cmake/compile_tests/get_sve_hw_vl.cpp
+      COMPILE_DEFINITIONS ${FLAG}
+      COMPILE_OUTPUT_VARIABLE COMPILE_OUTPUT_GET_SVE_HW_VL
+      RUN_OUTPUT_VARIABLE SVE_HW_VL
+    )
 
     if(RUN_GET_SVE_HW_VL EQUAL 0)
       # match the output SVE_HW_VL=<VL>
       string(REGEX MATCH "SVE_HW_VL=([0-9]+)" SVE_HW_VL "${SVE_HW_VL}")
       # remove "SVE_HW_VL="
-     string(REPLACE     "SVE_HW_VL=" ""      SVE_HW_VL "${SVE_HW_VL}")
+      string(REPLACE "SVE_HW_VL=" "" SVE_HW_VL "${SVE_HW_VL}")
       message(STATUS "Performing Test GET_SVE_HW_VL = ${SVE_HW_VL} -- success")
     else()
       if(NOT COMPILE_GET_SVE_HW_VL)
-        message(WARNING "Performing Test GET_SVE_HW_VL -- failed to compile with flag ${FLAG}: ${COMPILE_OUTPUT_GET_SVE_HW_VL}")
+        message(
+          WARNING
+            "Performing Test GET_SVE_HW_VL -- failed to compile with flag ${FLAG}: ${COMPILE_OUTPUT_GET_SVE_HW_VL}"
+        )
       else()
         message(WARNING "Performing Test GET_SVE_HW_VL -- compiled with flag ${FLAG} but failed to run: ${SVE_HW_VL}")
       endif()
@@ -385,14 +389,11 @@ if(KOKKOS_ARCH_ARMV84_SVE)
   set(KOKKOS_ARCH_ARMV84_SVE_FLAG -march=armv8.4-a+sve)
   check_cxx_compiler_flag(${KOKKOS_ARCH_ARMV84_SVE_FLAG} COMPILER_SUPPORTS_ARMV84_SVE)
 
-  if (COMPILER_SUPPORTS_ARMV84_SVE)
+  if(COMPILER_SUPPORTS_ARMV84_SVE)
     set(KOKKOS_ARCH_ARM_SVE ON)
-    GET_SVE_HW_VL(${KOKKOS_ARCH_ARMV84_SVE_FLAG})
+    get_sve_hw_vl(${KOKKOS_ARCH_ARMV84_SVE_FLAG})
     set(KOKKOS_ARCH_ARMV84_SVE_FLAG ${KOKKOS_ARCH_ARMV84_SVE_FLAG};-msve-vector-bits=${SVE_HW_VL})
-    compiler_specific_flags(
-      COMPILER_ID KOKKOS_CXX_HOST_COMPILER_ID
-      DEFAULT ${KOKKOS_ARCH_ARMV84_SVE_FLAG}
-    )
+    compiler_specific_flags(COMPILER_ID KOKKOS_CXX_HOST_COMPILER_ID DEFAULT ${KOKKOS_ARCH_ARMV84_SVE_FLAG})
   else()
     message(WARNING "Compiler does not support ARMv8.4-a+SVE architecture")
   endif()
@@ -771,[13](https://github.com/kokkos/kokkos/actions/runs/14127124111/job/39727674445?pr=7807#step:4:14) +772,22 @@ if(KOKKOS_ARCH_NATIVE)
   elseif(KOKKOS_COMPILER_HAS_ARM_SVE)
     message(STATUS "SIMD: ARM_SVE detected")
     set(KOKKOS_ARCH_ARM_SVE ON)
-    GET_SVE_HW_VL("${KOKKOS_NATIVE_FLAGS}")
+    get_sve_hw_vl("${KOKKOS_NATIVE_FLAGS}")
     compiler_specific_flags(
-      COMPILER_ID KOKKOS_CXX_HOST_COMPILER_ID
-      Clang   ${KOKKOS_NATIVE_FLAGS} -msve-vector-bits=${SVE_HW_VL}
-      GNU     ${KOKKOS_NATIVE_FLAGS} -msve-vector-bits=${SVE_HW_VL}
-      NVHPC   ${KOKKOS_NATIVE_FLAGS} -msve-vector-bits=${SVE_HW_VL}
-      DEFAULT ${KOKKOS_NATIVE_FLAGS} -msve-vector-bits=${SVE_HW_VL}
+      COMPILER_ID
+      KOKKOS_CXX_HOST_COMPILER_ID
+      Clang
+      ${KOKKOS_NATIVE_FLAGS}
+      -msve-vector-bits=${SVE_HW_VL}
+      GNU
+      ${KOKKOS_NATIVE_FLAGS}
+      -msve-vector-bits=${SVE_HW_VL}
+      NVHPC
+      ${KOKKOS_NATIVE_FLAGS}
+      -msve-vector-bits=${SVE_HW_VL}
+      DEFAULT
+      ${KOKKOS_NATIVE_FLAGS}
+      -msve-vector-bits=${SVE_HW_VL}
     )
```

</details>

hominhquan commented:
The clang-format issue was fixed 3 days ago, on the latest commit `d1120f2`
![image](https://github.com/user-attachments/assets/3ab49a40-d210-42a8-8c25-f3ec44517d67)

hominhquan commented:
@ldh4 After SNL CI, I added a new commit 77d056ef7f57f2f6c91af48b79d9f43688eb6546, which add explicit `basic_simd_mask<>` instantiation on 128-bit supported type and vector lanes `<T, abi>`, due to weird template error on the `is_type_v<>` boolean in the SIMD unit tests.

ldh4 commented:
Retest this please.

hominhquan commented:
> Retest this please.

Is this message for me ? if yes, how do you expect me to proceed ? (I don't have right to relaunch SNL-CI jobs)

ldh4 commented:
> > Retest this please.
> 
> Is this message for me ? if yes, how do you expect me to proceed ? (I don't have right to relaunch SNL-CI jobs)

No, that's the command we use to retrigger the testings.

masterleinad commented on simd/src/Kokkos_SIMD_SVE.hpp:
```diff
@@ -0,0 +1,3175 @@
+// NOTE: We observed that LLVM-14 and LLVM-15 defined __ARM_NEON_SVE_BRIDGE,
+// but not provided arm_neon_sve_bridge.h (!?)
+// https://godbolt.org/z/7jqWvzvWY
+KOKKOS_IMPL_WARNING(
+    "[WARNING]: You are using a compiler with __ARM_NEON_SVE_BRIDGE set, but \
+the header arm_neon_sve_bridge.h is missing. This could be a compiler \
+toolchain issue (observed on LLVM-14 and LLVM-15, see \
+https://godbolt.org/z/7jqWvzvWY). Kokkos will use its own implementation of \
+these functions, which could be slower than native ones in a recent SVE \
+compiler.")
```
This looks really annoying. How would a user disable the warning after they acknowledged it's contents?

hominhquan replied:
The warning is managed by the existing macro `KOKKOS_IMPL_WARNING`. Unless Kokkos provide a yet-exist `KOKKOS_DISABLE_WARNING`, we can not hide them at compile time.

tretre91 commented on Makefile.kokkos:
```diff
@@ -427,11 +427,13 @@ endif
 # ARM based.
 KOKKOS_INTERNAL_USE_ARCH_ARMV80 := $(call kokkos_has_string,$(KOKKOS_ARCH),ARMv80)
 KOKKOS_INTERNAL_USE_ARCH_ARMV81 := $(call kokkos_has_string,$(KOKKOS_ARCH),ARMv81)
+KOKKOS_INTERNAL_USE_ARCH_ARMV84 := $(call kokkos_has_string,$(KOKKOS_ARCH),ARMV84)
```
```suggestion
KOKKOS_INTERNAL_USE_ARCH_ARMV84 := $(call kokkos_has_string,$(KOKKOS_ARCH),ARMv84)
```
Just for consistency

tretre91 commented on simd/src/Kokkos_SIMD_SVE.hpp:
```diff
@@ -0,0 +1,3175 @@
+      std::int32_t,
+      Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>> const& a) {
+  return Experimental::basic_simd<
+      std::int32_t,
+      Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>>(
+      static_cast<vls_int32_t>(a));
+}
+
+[[nodiscard]] KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION Experimental::basic_simd<
+    std::int32_t, Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>>
+ceil(Experimental::basic_simd<
+     std::int32_t,
+     Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>> const& a) {
+  return Experimental::basic_simd<
+      std::int32_t,
+      Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>>(
+      static_cast<vls_int32_t>(a));
+}
+
+[[nodiscard]] KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION Experimental::basic_simd<
+    std::int32_t, Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>>
+round(Experimental::basic_simd<
+      std::int32_t,
+      Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>> const& a) {
+  return Experimental::basic_simd<
+      std::int32_t,
+      Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>>(
+      static_cast<vls_int32_t>(a));
+}
+
+[[nodiscard]] KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION Experimental::basic_simd<
+    std::int32_t, Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>>
+trunc(Experimental::basic_simd<
+      std::int32_t,
+      Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>> const& a) {
+  return Experimental::basic_simd<
+      std::int32_t,
+      Experimental::simd_abi::sve_fixed_size<SVE_WORDS_IN_VECTOR>>(
+      static_cast<vls_int32_t>(a));
+}
```
@ldh4 Should these functions return a `basic_simd<double, ...>` instead? Currently, the scalar, avx2 and avx512 backends return double, but the neon backend returns an integral type, I don't know what is the desirable behavior here.

If we should return an integral type, I think we can just do `return a;` instead of casting to the `implementation_type` and back to `basic_simd`
(same remark for the other integral type specializations)

ldh4 replied:
> Should these functions return a basic_simd<double, ...> instead? Currently, the scalar, avx2 and avx512 backends return double, but the neon backend returns an integral type, I don't know what is the desirable behavior here.

That may have been an oversight on neon backend impl. They should all be returning `basic_simd<double...>` for floating point math functions. However, I think this correction could be handled in a follow up PR along with corrections in neon backend.

tretre91 commented on simd/src/Kokkos_SIMD_SVE.hpp:
```diff
@@ -0,0 +1,3175 @@
+template <class Derived, int Bits>
+class sve_mask;
+
+template <int nbits, int SVE_T_IN_VECTOR>
+KOKKOS_IMPL_HOST_FORCEINLINE_FUNCTION static vls_bool_t get_pred(
+    std::size_t lane) noexcept {
+  if constexpr (nbits == 8) {
+    return svwhilele_b8(0, static_cast<std::int32_t>(lane));
+  } else if constexpr (nbits == 16) {
+    return svwhilele_b16(0, static_cast<std::int32_t>(lane));
+  } else if constexpr (nbits == 32) {
+    return svwhilele_b32(0, static_cast<std::int32_t>(lane));
+  } else if constexpr (nbits == 64) {
+    return svwhilele_b64(0, static_cast<std::int32_t>(lane));
+  } else {
+    __builtin_unreachable();
+  }
+}
```
Is the template parameter `SVE_T_IN_VECTOR` required?

hominhquan replied:
removed SVE_T_IN_VECTOR

tretre91 reviewed (changes requested):
Is there a specific reason why  `is_simd_avail_v` is not used in place of `is_type_v` in device tests?

dssgabriel commented:
Is there a specific reason why the predication mode is not consistent across all the provided functions/operators?

**Context for other reviewers who may not be familiar with Arm SVE**
SVE has two "masking policies" when executing predicated instructions:
- **merging**: Inactive elements in the destination vector register remain unmodified;
- **zeroing**: Inactive elements in the destination vector register are set to zero (equivalent to merging with a zeroed vector register).

I noticed that some operations are implemented using zeroing predication (intrinsics suffixed with `_z`), while others are implemented using the ACLE-specific "don't care" predication (intrinsics suffixed with `_x`), where the inactive elements in the destination vector register are set to an arbitrary/undefined value. Note that this predication mode has no equivalent in the AArch64 ISA; only merging and zeroing predication exist.

Is this inconsistency in the implementation a desired behavior? I know that some instructions may exhibit a performance penalty when using zeroing mode (depending on the microarchitecture). If that is the case, I think this choice should be detailed in explanatory comments, as it is not immediately obvious.
Otherwise, I think it is better to use a single predication mode consistently (preferably zeroing, as it is deterministic and maps to an actual feature of the ISA).

hominhquan commented:
> Is there a specific reason why `is_simd_avail_v` is not used in place of `is_type_v` in device tests?

It could/should be applied to device as well. But there are reasons I did not do it:
- SIMD backend is today mostly focusing on CPU (x86 and aarch64)
- Device 's `basic_simd` type is currently implemented as [scalar (including CUDA, SYCL and OPENMPTARGET) ](https://github.com/kokkos/kokkos/blob/develop/simd/src/Kokkos_SIMD.hpp#L126)
- and any device-related support is not in the scope of this PR, as long as the device part passed all the CI.

hominhquan commented:
@dssgabriel Your comment is relevant. I admit that I kept these modes from the original commit of @ldh4. Now given that the SIMD API does not support masked/predicated-arithmetic, all operations are performed with the all-true (`svptrue*`) SVE predicate, so we can pick a unique mode for all. @ldh4 and @dssgabriel what is your opinion about this ?

ldh4 reviewed (commented):
> @dssgabriel Your comment is relevant. I admit that I kept these modes from the original commit of @ldh4. Now given that the SIMD API does not support masked/predicated-arithmetic, yet, all operations are performed with the all-true (svptrue*) SVE predicate, so we can pick a unique mode for all. @ldh4 and @dssgabriel what is your opinion about this ? should we pick _m, _x or _z ?

The standard notes using `T()` for masked inits and loads. For currently supported simd data types in Kokkos, `_z` should be suitable for masked inits and loads. For all other masked operations, I think `_m` should be used.

tretre91 commented:
> > Is there a specific reason why `is_simd_avail_v` is not used in place of `is_type_v` in device tests?
> 
> It could/should be applied to device as well. But there are reasons I did not do it:
> 
>   * SIMD backend is today mostly focusing on CPU (x86 and aarch64)
> 
>   * Device 's `basic_simd` type is currently implemented as [scalar (including CUDA, SYCL and OPENMPTARGET) ](https://github.com/kokkos/kokkos/blob/develop/simd/src/Kokkos_SIMD.hpp#L126)
> 
>   * and any device-related support is not in the scope of this PR, as long as the device part passed all the CI.

Ok, just wanted to be sure. These changes can be applied in another PR

hominhquan commented:
I'll apply these new requested changes when I'll be back from travel.

hominhquan commented:
@tretre91 @dssgabriel @ldh4 I added a new commit 0cfcec0 to this PR 

- basic_simd arithmetic operations to use _m suffix (except sve_mask), since only `_z` is allowed for OR, AND, XOR and NOT between SVE masks/predicates (SVE doc)

- After reflection, I decide to keep `#if __has_include(<arm_neon_sve_bridge.h>)` to include `arm_neon_sve_bridge.h` instead of checking for a new KOKKOS macro, and just add a warning message in CMake configure. The reason to this is: 
  - The current SIMD backend is header-only (`simd/src/Kokkos_SIMD_[NEON,SVE,AVX2,AVX52].hpp`), i.e. will be compiled at user side.
  - A possible situation would be Kokkos's core runtime was built with a compiler version x.x (without `arm_neon_sve_bridge.h` => a macro `KOKKOS_MISSING_ARM_NEON_SVE_BRIDGE` set to 1 somewhere in `KokkosCore_config.h`), then delivered to another user who will use a compiler y.y, which now has `arm_neon_sve_bridge.h`, but can not use it since the macro  `KOKKOS_MISSING_ARM_NEON_SVE_BRIDGE`  was set. It would be regrettable to ignore the native Neon-SVE bridge by the compiler y.y just because Kokkos was configured with the compiler x.x.
  - So, using `__has_include(<arm_neon_sve_bridge.h>)` in `Kokkos_SIMD_SVE.hpp` will work seamlessly in the both cases of missing or presence of `arm_neon_sve_bridge.h`, even in the case of mismatch of compiler version.

- Feel free to tell me if this commit should be squashed into a previous commit.

tretre91 reviewed (approved):
Looks good to me

hominhquan commented:
Retest this please.

hominhquan commented:
> Retest this please.

@ldh4 it looks like the magic sentence does not work to me :)

tretre91 commented:
> > Retest this please.
> 
> @ldh4 it looks like the magic sentence does not work to me :)

The workflows needed manual approval from a maintainer for some reason

ldh4 commented:
Please run `cmake-format` on `CMakeLists.txt` again.
```
--- a/simd/src/CMakeLists.txt
+++ b/simd/src/CMakeLists.txt
@@ -26,16 +26,17 @@ kokkos_lib_include_directories(
 # If ARM-SVE, check and warn for available arm_neon_sve_bridge.h
 if(KOKKOS_ARCH_ARM_SVE)
   check_include_file_cxx(arm_neon_sve_bridge.h KOKKOS_COMPILER_SUPPORTS_ARM_NEON_SVE_BRIDGE)
-  if (NOT KOKKOS_COMPILER_SUPPORTS_ARM_NEON_SVE_BRIDGE)
-    message(WARNING "You are using a compiler without NEON-SVE bridge header "
-                    "(arm_neon_sve_bridge.h). Kokkos will use its own "
-                    "implementation of these functions, which could be slower "
-                    "than native ones from a more recent SVE compiler. It is "
-                    "recommended to upgrade your compiler.\n"
-                    "NOTE: It was also observed that, in some LLVM versions "
-                    "(14, 15), despite the macro __ARM_NEON_SVE_BRIDGE was "
-                    "defined, the header arm_neon_sve_bridge.h is still not "
-                    "available (see https://godbolt.org/z/7jqWvzvWY )."
+  if(NOT KOKKOS_COMPILER_SUPPORTS_ARM_NEON_SVE_BRIDGE)
+    message(
+      WARNING "You are using a compiler without NEON-SVE bridge header "
+              "(arm_neon_sve_bridge.h). Kokkos will use its own "
+              "implementation of these functions, which could be slower "
+              "than native ones from a more recent SVE compiler. It is "
+              "recommended to upgrade your compiler.\n"
+              "NOTE: It was also observed that, in some LLVM versions "
+              "(14, 15), despite the macro __ARM_NEON_SVE_BRIDGE was "
+              "defined, the header arm_neon_sve_bridge.h is still not "
+              "available (see https://godbolt.org/z/7jqWvzvWY )."
     )
```

hominhquan commented:
@ldh4 retrigger please

ldh4 commented:
Retest this please.

ldh4 commented:
Retest this please.

hominhquan commented:
We still have issues with some JAVA errors in Jenkins `clang-format` and NVIDIA-GH200 unavailable.

ldh4 commented:
Retest this please.

ldh4 commented:
Failures in the CI seem unrelated to the changes in this PR.

```
53/54 Test #53: Kokkos_UnitTest_SIMD .......................................   Passed    0.27 sec
test 54
      Start 54: Kokkos_Example_RelocatableFunction
54: Test command: /var/jenkins/workspace/Kokkos_PR-7807/build-tests/example/relocatable_function/Kokkos_example_relocatable_function
54: Test timeout computed to be: 1500
54: 
54: Kokkos::Cuda::initialize WARNING: Cuda is allocating into UVMSpace by default
54:                                   without setting CUDA_MANAGED_FORCE_DEVICE_ALLOC=1 or
54:                                   setting CUDA_VISIBLE_DEVICES.
54:                                   This could on multi GPU systems lead to severe performance"
54:                                   penalties.
54: 1.000000e+01 2.985000e-05s (relocatable) vs. 1.587000e-05s (inline)
54: 1.000000e+02 1.487000e-05s (relocatable) vs. 1.482000e-05s (inline)
54: 1.000000e+03 1.467000e-05s (relocatable) vs. 1.483000e-05s (inline)
54: 1.000000e+04 1.474100e-05s (relocatable) vs. 1.479000e-05s (inline)
54: 1.000000e+05 2.000000e-05s (relocatable) vs. 1.916000e-05s (inline)
54: 1.000000e+06 2.282100e-05s (relocatable) vs. 2.089000e-05s (inline)
54: 1.000000e+07 4.777000e-05s (relocatable) vs. 3.073100e-05s (inline)
54: 1.000000e+08 3.033650e-04s (relocatable) vs. 1.414920e-04s (inline)
54/54 Test #54: Kokkos_Example_RelocatableFunction .........................   Passed    0.24 sec
98% tests passed, 1 tests failed out of 54
Label Time Summary:
Kokkos    = 352.81 sec*proc (52 tests)
Total Test time (real) = 357.07 sec
The following tests FAILED:
	 34 - Kokkos_UnitTest_Random (Failed)
Errors while running CTest
script returned exit code 8
```

dssgabriel commented:
Congrats on merging this PR, @hominhquan! :partying_face:

hominhquan commented:
Big thank to everyone involved in this PR for your time and contribution.
